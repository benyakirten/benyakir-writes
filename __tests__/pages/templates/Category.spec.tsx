import React from "react";
import { cleanup, render, screen, fireEvent, act } from "@testing-library/react";

import CategoryTemplate from "@/templates/Category.template";
import { WpPostByCategory } from "@Types/query";

describe("category template", () => {
    const testData: WpPostByCategory[] = [
        {
            data: {
                allWpPost: {
                    nodes: [],
                },
            },
            pageContext: {
                name: "category A",
            },
        },
        {
            data: {
                allWpPost: {
                    nodes: [
                        {
                            title: "blog post A",
                            slug: "blog-post-a",
                            excerpt: "blog post A excerpt",
                            date: "2019/09/15",
                            content: "blog post A content",
                            categories: { nodes: null },
                            tags: { nodes: null },
                        },
                        {
                            title: "blog post B",
                            slug: "blog-post-b",
                            excerpt: "blog post B excerpt",
                            date: "2019/10/15",
                            content: "blog post B content",
                            categories: { nodes: null },
                            tags: { nodes: null },
                        },
                    ],
                },
            },
            pageContext: {
                name: "category B",
            },
        },
    ];

    beforeEach(() => {
        jest.useFakeTimers()
    })

    afterEach(() => {
        jest.runOnlyPendingTimers()
        jest.useRealTimers()
        cleanup()
    })

    it("should render correctly", () => {
        expect(() => render(<CategoryTemplate {...testData[0]} />)).not.toThrow()
        cleanup()
        expect(() => render(<CategoryTemplate {...testData[1]} />)).not.toThrow()
    });

    it('should render a boilerplate if there are no blog posts for the category', async () => {
        render(<CategoryTemplate {...testData[0]} />)

        const title = await screen.findByText('No posts exist')
        expect(title).toBeTruthy()
        expect(title.tagName).toEqual('H1')

        const para = title.nextElementSibling!
        expect(para.tagName).toEqual("P")
        expect(para.textContent).toEqual("For the category category A, at least. Maybe you want to check out the general blog page instead?")

        expect(para.firstElementChild?.getAttribute('href')).toEqual("/blog")
    })

    it('should render a header with the category name if it exists', async () => {
        render(<CategoryTemplate {...testData[1]} />)

        const title = await screen.findByText('category B')
        expect(title.tagName).toEqual('H1')

        const posts = await screen.findAllByRole('article')
        expect(posts.length).toEqual(2)
    })

    it('should render only the filtered items if there is a filter applied', async () => {
        render(<CategoryTemplate {...testData[1]} />)

        await act(async () => {
            const input = await screen.findByRole("textbox")
            fireEvent.change(input, { target: { value: 'september'} })

            jest.runAllTimers()
            
            const posts = await screen.findAllByRole('article')
            expect(posts.length).toEqual(1)
        })
    })
});
