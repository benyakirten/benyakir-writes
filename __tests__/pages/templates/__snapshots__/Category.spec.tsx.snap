// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`category template should render correctly 1`] = `
<div
  className="sc-jJEKmz dPuFhU"
>
  <h1
    className="sc-jSgvzq FRpcC"
  >
    Category A
  </h1>
  <div
    className="sc-gWHigU eRSuBn"
  >
    <span
      className="sc-jcVcSv dSigqm"
      size="2rem"
    />
  </div>
  <div
    className="sc-hiSbEG hObFLk"
  >
    <span
      className="sc-jcVcSv coWmFG"
      size="4rem"
    />
  </div>
</div>
`;

exports[`category template should render correctly 2`] = `
<div
  className="sc-jJEKmz dPuFhU"
>
  <h1
    className="sc-jSgvzq FRpcC"
  >
    Category B
  </h1>
  <div
    className="sc-gWHigU eRSuBn"
  >
    <span
      className="sc-jcVcSv dSigqm"
      size="2rem"
    />
  </div>
  <div
    className="sc-hiSbEG hObFLk"
  >
    <span
      className="sc-jcVcSv coWmFG"
      size="4rem"
    />
  </div>
</div>
`;
